rock-n-rollup/src/core/runtime.rs

```diff
@@ -104,132 +43,126 @@ pub trait Runtime: 'static {
-            let num_bytes = data.len();
-            let src = data.as_ptr();
-            store_write(path, path_len, at_offset, src, num_bytes)
-        };
+        let path = OwnedPath::try_from(path.to_string()).map_err(|_| ())?;
+
+        let res = self.host.store_write(&path, data, at_offset);
         match res {
-            0 => Ok(()),
-            err => {
-                self.write_debug(&format!("error store_write_raw: {}\n", err));
+            Ok(_) => Ok(()),
+            Err(_) => {
+                self.host.write_debug("Error store_write");
                 Err(())
             }
         }
     }
-    fn reveal_preimage(&mut self, hash: &[u8; PREIMAGE_HASH_SIZE]) -> Result<Vec<u8>, ()> {
-        let max_size = 4096;
-        let mut payload = Vec::with_capacity(MAX_MESSAGE_SIZE as usize);
 
-        let u8_size = u8::try_from(PREIMAGE_HASH_SIZE).unwrap();
+    fn store_move(&mut self, from: &str, to: &str) -> Result<(), ()> {
+        let from = OwnedPath::try_from(from.to_string()).map_err(|_| ())?;
+        let to = OwnedPath::try_from(to.to_string()).map_err(|_| ())?;
 
-        unsafe {
-            let size = reveal_preimage(hash.as_ptr(), u8_size, payload.as_mut_ptr(), max_size);
-            if size < 0 {
+        let res = self.host.store_move(&from, &to);
+        match res {
+            Ok(_) => Ok(()),
+            Err(_) => {
+                self.host.write_debug("Error store_move");
                 Err(())
-            } else {
-                let size = usize::try_from(size).unwrap();
-                payload.set_len(size);
-                Ok(payload)
             }
         }
     }
 
-    fn store_move(&mut self, from: &str, to: &str) -> Result<(), ()> {
-        let res = unsafe { store_move(from.as_ptr(), from.len(), to.as_ptr(), to.len()) };
+    fn reveal_preimage(&mut self, hash: &[u8; PREIMAGE_HASH_SIZE]) -> Result<Vec<u8>, ()> {
+        let payload: Vec<u8> = Vec::with_capacity(MAX_MESSAGE_SIZE);
+        let mut payload: Vec<u8> = payload.to_vec();
+        let res: Result<usize, _> = self.host.reveal_preimage(hash, &mut payload);
         match res {
-            0 => Ok(()),
-            _ => Err(()),
+            Ok(bytes) => Ok(bytes.to_le_bytes().to_vec()),
+            Err(_) => {
+                self.host.write_debug("Error reveal_preimage");
+                Err(())
+            }
+        }
+    }
+
+    fn store_is_present(&mut self, path: &str) -> bool {
+        let path = OwnedPath::try_from(path.to_string()).map_err(|_| ());
+        match path {
+            Ok(path) => {
+                let res = { self.host.store_has(&path) };
+                match res {
+                    Ok(Some(_value_type)) => true,
+                    Ok(None) => false, // no file
+                    Err(_) => false,
+                }
+            }
+            Err(_) => false,
+        }
+    }
+
+    fn next_input(&mut self) -> Option<RawInput> {
+        let size = self.host.read_input();
```

Comment:
size seems to be an artefact from the old version, a new name might be more appropriate

Reply:
I rename it to `msg_opt` at this [commit](3d56b448ce27ff2ee25b9e46e3b86aecbbb74896)